### PR TITLE
Add option to add custom buttons to tree toolbar

### LINKED ISSUE #373: Option to control what buttons are available in tree toolbars

### PR DESCRIPTION
**Pull request by simasjar**
Made the option to add custom buttons to the tree toolbar

For example, if the user wants to show only the "Show all" button in Model Tree:

```tsx
UiItemsManager.register(new TreeWidgetUiItemsProvider({
  modelsTreeProps: {
    headerButtons: [
      (props) => <ModelsTreeComponent.ShowAllButton {...props} />
    ]
  }
}));
```

Also fixed a bug where header buttons would overlay with other buttons:
before:
image
after:
image
image
when the search button is pressed:
image

closes #373

### CODE REVIEW

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 1]
```diff
@@ -69,106 +67,186 @@ export const ModelsTreeComponent = (props: ModelTreeProps) => {
...
+  );
+};
 
-  const hideAll = useCallback(async () => {
-    viewport?.changeModelDisplay(availableModels, false);
-    viewport?.clearAlwaysDrawn();
-    if (iModel) {
+export function ShowAllButtonModelsTree(props: ModelsTreeHeaderButtonProps) {
+  const showAll = useCallback(async () => {
+    if (!props.availableModels)
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:108]: The suggestion applies to all the buttons

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx [hunk 2]
```diff
@@ -27,38 +29,99 @@ export function CategoriesTreeComponent(props: CategoriesTreeProps) {
...
+      {iModel && viewport &&
+        <>
+          <TreeHeaderComponent
+            searchOptions={searchOptions}
+            treeHeaderButtons={props.TreeHeaderButtons
+              ? props.TreeHeaderButtons.map((btn) => btn({ iModel, viewport, filteredProvider }))
+              : [
+                <ShowAllButtonCategoriesTree iModel={iModel} filteredProvider={filteredProvider} key="show-all-btn" />,
+                <HideAllButtonCategoriesTree iModel={iModel} filteredProvider={filteredProvider} key="hide-all-btn" />,
+                <InvertButtonCategoriesTree iModel={iModel} viewport={viewport} filteredProvider={filteredProvider} key="invert-all-btn" />,
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx:43]: Maybe move all the buttons under `CategoriesTreeComponent` namespace?

File: packages/itwin/tree-widget/src/types.ts [hunk 3]
```diff
@@ -5,34 +5,46 @@
...
 import type { SelectableContentDefinition } from "@itwin/components-react";
-import type { IModelConnection } from "@itwin/core-frontend";
+import type { IModelConnection, ScreenViewport } from "@itwin/core-frontend";
+import { IPresentationTreeDataProvider } from "@itwin/presentation-components";
 import type { CategoryTreeProps } from "./components/trees/category-tree/CategoriesTree";
 import type { ModelsTreeProps } from "./components/trees/models-tree/ModelsTree";
 
+export interface TreeHeaderButtonProps {
+  viewport?: ScreenViewport;
+  iModel?: IModelConnection;
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/itwin/tree-widget/src/types.ts:15]: also, could we include the tree model here? we may need that to get selected nodes, etc.

[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/types.ts:15]: In order to include `model` in button props a lot of refactoring is required to move all the state from `ModelsTree` to `ModelsTreeComponent` where header and `ModelsTree` is renderer. I think this could be done in a separate PR.

File: packages/itwin/tree-widget/src/components/header/TreeHeader.tsx [hunk 4]
```diff
@@ -3,32 +3,19 @@
...
 export interface TreeHeaderComponentProps {
-  showAll?: () => void;
-  hideAll?: () => void;
-  invert?: () => void;
-  toggle2D?: () => void;
-  toggle2DIcon?: string;
-  toggle3D?: () => void;
-  toggle3DIcon?: string;
   searchOptions: SearchOptions;
+  treeHeaderButtons: React.ReactNode[];
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/header/TreeHeader.tsx:13]: I thinks after changes to `SearchBar` `TreeHeaderComponent` can also take buttons as `children` instead of taking them through props

File: packages/itwin/tree-widget/src/types.ts [hunk 5]
```diff
@@ -5,34 +5,46 @@
...
-ModelsTreeProps,
+export interface ModelTreeProps extends Omit<ModelsTreeProps,
 | "iModel"
 | "activeView"
 | "width"
 | "height"
 | "filterInfo"
 | "onFilterApplied"
->;
+> { TreeHeaderButtons?: Array<(props: ModelsTreeHeaderButtonProps) => React.ReactNode> }
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/types.ts:38]: Property name should start with lower case letter

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 6]
```diff
@@ -69,106 +67,184 @@ export const ModelsTreeComponent = (props: ModelTreeProps) => {
...
+    </>
+  );
+};
 
-  const hideAll = useCallback(async () => {
-    viewport?.changeModelDisplay(availableModels, false);
-    viewport?.clearAlwaysDrawn();
-    if (iModel) {
+export function ShowAllButtonModelsTree(props: ModelsTreeHeaderButtonProps) {
+  const showAll = useCallback(async () => {
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:107]: I think there is no need to wrap `onClick` handler with `useCallback` in all of these buttons

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 7]
```diff
@@ -69,106 +67,184 @@ export const ModelsTreeComponent = (props: ModelTreeProps) => {
...
-    viewport?.invalidateScene();
-  }, [viewport, availableModels]);
+  return (
+    <>
+      {iModel && viewport &&
+        <>
+          <TreeHeaderComponent
+            searchOptions={searchOptions}
+            treeHeaderButtons={props.TreeHeaderButtons
+              ? props.TreeHeaderButtons.map((btn) => btn({ viewport, iModel, availableModels }))
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:77]: React logs an error in console for missing `key` property.

File: common/scripts/install-run-rush.js [hunk 8]
```diff
@@ -1,3 +1,7 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (aruniverse) [inline common/scripts/install-run-rush.js:1]: could you revert the changes in here, and commit using `no-verify` to prevent addition of bentley copyright headers?

same thing for the install-run.js script

File: packages/itwin/tree-widget/package.json [hunk 9]
```diff
@@ -41,7 +41,9 @@
     "i18next": "^10.2.2",
     "react-redux": "^7.2.0",
     "redux": "^4.0.3",
-    "rxjs": "^6.6.2"
+    "rxjs": "^6.6.2",
+    "@itwin/itwinui-react": "^1.40.1",
```
Comment (aruniverse) [inline packages/itwin/tree-widget/package.json:45]: alphabetize plz

File: common/scripts/install-run-rush.js (hunk 8 above)
Comment (simasjar) [reply common/scripts/install-run-rush.js]: fixed

File: packages/itwin/tree-widget/package.json (hunk 9 above)
Comment (simasjar) [reply packages/itwin/tree-widget/package.json]: fixed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 10]
```diff
@@ -69,106 +67,184 @@ export const ModelsTreeComponent = (props: ModelTreeProps) => {
...
+      icon={icon2dToggle}
+      title={TreeWidget.translate("toggle2DViews")}
+      onClick={viewToggle2D}
+      label={TreeWidget.translate("label2D")}
+    />
+  );
+}
+
+export function View3DButtonModelsTree(props: ModelsTreeHeaderButtonProps) {
+  const [is3dToggleActive, setIs3dToggleActive] = useState<boolean>(false);
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:222]: These state variables probably will reset when button is inside dropdown menu (Searbar open) and dropdown menu is reopened

**Review by saskliutas — CHANGES_REQUESTED** [review]
Is Popup component still used? Maybe it can be removed?

**Comment by grigasp** [comment]
> Please take a look. We are trying to decouple the pkg from being tightly coupled with DR; so you will most likely have to make changes on your own

Just to add to this. The goal here is to allow controlling what buttons are displayed in the toolbar and supply custom ones. The default list of buttons, at least at this point, does not change.

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: Made viewport as a required parameter in TreeHeaderButtonProps

File: packages/itwin/tree-widget/src/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: fixed

File: packages/itwin/tree-widget/src/types.ts [hunk 11]
```diff
@@ -5,34 +5,46 @@
...
+  viewport?: ScreenViewport;
+  iModel?: IModelConnection;
+}
+
+export interface ModelsTreeHeaderButtonProps extends TreeHeaderButtonProps {
+  availableModels?: string[];
+}
+
+export interface CategoriesTreeHeaderButtonProps extends TreeHeaderButtonProps {
+  filteredProvider?: IPresentationTreeDataProvider;
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: fixed

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx (hunk 2 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/components/header/TreeHeader.tsx (hunk 4 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/header/TreeHeader.tsx]: fixed

File: packages/itwin/tree-widget/src/types.ts (hunk 5 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: fixed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 6 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 7 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 10 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 12]
```diff
@@ -4,171 +4,255 @@
...
 import { toggleAllCategories } from "../CategoriesVisibilityUtils";
 import { ModelsTree } from "./ModelsTree";
+import { IconButton } from "../../IconButton";
+import { TreeWidget } from "../../../TreeWidget";
 
-interface TreeViewModelInfo {
-  id: string;
-  isPlanProjection?: boolean;
-}
+export namespace ModelsTreeComponentNamespace {
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:20]: I would suggest instead of using `namespace` put buttons on `ModelsTreeComponent`:

```typescript
export function ModelsTreeComponent(props: ModelTreeProps) {
  // models tree component implementation
}

ModelsTreeComponent.ShowAllButton = ShowAllButton;
ModelsTreeComponent.HideAllButton = HideAllButton;
ModelsTreeComponent.InvertButton = InvertButton;
ModelsTreeComponent.View2DButton = View2DButton;
ModelsTreeComponent.View3DButton = View3DButton;
```

And then there is no need to export each button component itself. Same for `CategoriesTreeComponent`

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 13]
```diff
@@ -4,171 +4,255 @@
...
+      />
+    );
+  }
+
+  export function View2DButton(props: ModelsTreeHeaderButtonProps) {
+    const models2d = useMemo(() => {
+      return props.models?.filter((model) => model.isPlanProjection).map((model) => model.id);
+    }, [props.models]);
+
+    const [is2dToggleActive, setIs2dToggleActive] = useState(() => (models2d && models2d.length !== 0) ? models2d.every((id) => props.viewport.viewsModel(id)) : false);
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:195]: Move logic determining if toggle is active to separate function to avoid duplicating it in `useState` and `useEffect`

File: packages/itwin/tree-widget/.eslintrc.json [hunk 14]
```diff
@@ -21,6 +21,13 @@
           "internal"
         ]
       }
+    ],
+    "@typescript-eslint/no-floating-promises": [
```
Comment (saskliutas) [inline packages/itwin/tree-widget/.eslintrc.json:25]: Doesn't adding `void` solve this?

```typescript
void (async () => { /* some work*/ })();
```

File: packages/itwin/tree-widget/src/types.ts [hunk 15]
```diff
@@ -5,34 +5,46 @@
...
 import type { CategoryTreeProps } from "./components/trees/category-tree/CategoriesTree";
 import type { ModelsTreeProps } from "./components/trees/models-tree/ModelsTree";
+import type { Category, ModelsTreeComponentNamespace } from "./tree-widget-react";
+
+export interface TreeHeaderButtonProps {
+  viewport: ScreenViewport;
+}
+
+export interface ModelsTreeHeaderButtonProps extends TreeHeaderButtonProps {
+  models?: ModelsTreeComponentNamespace.TreeViewModelInfo[];
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/types.ts:18]: I think there is not need to put `TreeViewModelInfo` under namespace. Also it probebly could be renamed to `ModelInfo`?

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/types.ts [hunk 16]
```diff
@@ -35,7 +35,7 @@ export interface ModelTreeProps extends Omit<ModelsTreeProps,
 | "height"
 | "filterInfo"
 | "onFilterApplied"
-> { TreeHeaderButtons?: Array<(props: ModelsTreeHeaderButtonProps) => React.ReactNode> }
+> { treeHeaderButtons?: Array<(props: ModelsTreeHeaderButtonProps) => React.ReactNode> }
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/types.ts:38]: maybe change `treeHeaderButtons` to `headerButtons`? interface name already tells that's the attribute is related to a tree

File: packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx (hunk 2 above)
Comment (grigasp) [reply packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx]: My idea was to move just the buttons and name the namespace with the same name as the component (`CategoriesTreeComponent`). But I like @saskliutas suggestion even better: https://github.com/iTwin/viewer-components-react/pull/396#discussion_r1139900293.

Comment (grigasp) [reply packages/itwin/tree-widget/src/components/trees/category-tree/CategoriesTreeComponent.tsx]: My idea was to move only the buttons, and name the namespace similar to the component - `CategoriesTreeComponent`. But I like @saskliutas suggestion even better: https://github.com/iTwin/viewer-components-react/pull/396#discussion_r1139900293

File: packages/itwin/tree-widget/src/components/header/TreeHeader.tsx [hunk 17]
```diff
@@ -44,55 +30,7 @@ export function TreeHeaderComponent({
...
-            className={"tree-widget-header-tree-toolbar-icon"}
-            key="view-3d-btn"
-            icon={toggle3DIcon}
-            title={TreeWidget.translate("toggle3DViews")}
-            onClick={toggle3D}
-            label={TreeWidget.translate("label3D")}
-          />
-        )}
-      </div>
+      {children}
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/header/TreeHeader.tsx]: Removed it

File: packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss [hunk 18]
```diff
@@ -61,7 +64,6 @@
   .search-bar-search-container {
     flex: 1;
     height: 100%;
-    position: absolute;
     left: 0;
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss]: removed

File: packages/itwin/tree-widget/src/components/search-bar/SearchBar.tsx [hunk 19]
```diff
@@ -114,65 +109,79 @@ SearchBarState
...
+      showSearch && "contracted",
       alignment === Alignment.Right && "right"
     );
     const searchBarContainerClassName = classnames(
       "search-bar-search-container",
-      enableGrouping && "search-bar-grouping-enabled",
+      enableGrouping && "search-bar-grouping-enabled"
     );
 
+    const buttonGroup = (
```
[the remark this answers is not in the document]
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/search-bar/SearchBar.tsx]: removed

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 13 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/types.ts (hunk 15 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: fixed

File: packages/itwin/tree-widget/src/types.ts (hunk 16 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: fixed

File: packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss [hunk 20]
```diff
@@ -79,12 +67,12 @@
       &.show {
         visibility: visible;
         opacity: 1;
-        width: 100%;
+        width: calc(100% + 32px);
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss:70]: Why do you need to add `32px`?

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx [hunk 21]
```diff
@@ -4,171 +4,273 @@
...
+          />
+        )}
+      </AutoSizer>
+    </>
+  );
+}
+
+function ShowAllButton(props: ModelsTreeHeaderButtonProps) {
+  const showAll = async () => {
+    if (!props.models || !props.viewport)
```
Comment (saskliutas) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx:123]: Looking at `ModelsTreeComponentImpl` it looks like `models` and `viewport` are always defined.

Changing `ModelsTreeHeaderButtonProps` to have non-optional `models` propety would allow to remove quite some `undefined` checks `?` in button implementations.

**Review by saskliutas — CHANGES_REQUESTED** [review]

File: packages/itwin/tree-widget/src/types.ts [hunk 22]
```diff
@@ -6,21 +6,21 @@
...
 
 export interface ModelsTreeHeaderButtonProps extends TreeHeaderButtonProps {
-  availableModels?: string[];
+  models?: ModelInfo[];
 }
 
 export interface CategoriesTreeHeaderButtonProps extends TreeHeaderButtonProps {
-  filteredProvider?: IPresentationTreeDataProvider;
+  categories: Category[];
+  filteredCategories?: Category[];
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/types.ts:23]: Since we're making the `Category` type public, I think it needs some renaming:
- type to `CategoryInfo` to match what we have for models
- `key` attribute to `id` or `categoryId`
- `children` attribute to `subCategoryIds`

File: packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss (hunk 20 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/search-bar/SearchBar.scss]: The main idea was to overflow the dropdown button, but now the dropdown button won't be rendered if enableGrouping is set to false

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx (hunk 21 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx]: fixed

File: packages/itwin/tree-widget/src/types.ts (hunk 22 above)
Comment (simasjar) [reply packages/itwin/tree-widget/src/types.ts]: renamed

**Review by grigasp — APPROVED** [review]
LGTM. I think we need an additional change file to mention button behavior changes made in this PR.

**Review by aruniverse — APPROVED** [review]
